### ISSUE TITLE
Support pytrees directly in the API?

### ISSUE THREAD

**Issue by tjltjl** [pnkraemer/matfree#226]
I really like matfree, it is filling some important gaps in the Jax space.

There is one API improvement that would be really nice:

For background, please see [jax.scipy.sparse.linalg.cg](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html) - there, the interface directly allows pytrees for b and X0 and a function for A, which is *really* convenient (A^T is obtained via automatic differentiation).

I.e.:

1) all vectors are allowed to be generic pytrees,

2) all Av, vA function pairs can be defined by just providing Av and the code would just use jax.vjp on Av to obtain the vA function.

I'm enclosing a wrapper for eig.svd_partial following this convention, computing the highest singular values of a simple diffusion operator using a convolution. It's possible to inspect the eigenvectors after this code by simply doing
```
     imshow(comps[0][i])
```
for the i'th eigenvector.
![image](https://github.com/user-attachments/assets/28bf0788-030d-4d90-b2ac-dea657dc32a4)

There is only one difficulty: the axes of the returned u matrix are transposed to allow for the easy vmapping. I'm not sure what would be the most elegant way of handling this; having a structure of arrays with the last index of each array being the vmapped axis does not seem correct.

```
def svd_partial_simple(v0, depth, Av, vA=None):
    v0_flat, v_unravel = jax.flatten_util.ravel_pytree(v0)

    # We get this only after applying Av
    u_unravel = None

    def Av_flat(v_flat):
        result = Av(v_unravel(v_flat))
        nonlocal u_unravel
        result_flat, u_unravel = jax.flatten_util.ravel_pytree(result)
        return result_flat

    if vA:
        def vA_flat(u_flat):
            result = vA(u_unravel(u_flat))
            result_flat, _ = jax.flatten_util.ravel_pytree(result)
            return result_flat
    else:
        _, vA_flat_raw = jax.vjp(Av_flat, v0_flat)
        def vA_flat(u_flat):
            return vA_flat_raw(u_flat)[0]

    u, s, v = matfree.eig.svd_partial(v0_flat, depth, Av_flat, vA_flat)

    return (
        jax.vmap(u_unravel)(u.T),
        s,
        jax.vmap(v_unravel)(v),
    )

comps = svd_partial_simple(np.random.randn(5, 5), 10, 
        lambda v: 
            jax.lax.conv(v[None, None], jnp.ones((3, 3))[None, None], (1, 1), "SAME")[0, 0, :, :])
```

**Comment by pnkraemer** [comment]
Hi! Thanks for reaching out :blush:  

Do I understand that there are two separate features you'd like? 
1. Making `vA` optional and defaulting to using jax.linear_transpose to implement it?
2. Supporting pytrees in eig.svd_partial.


As for the first one, yes, I think that would be nice -- but it should probably be implemented on the level of `matfree.decomp.bidiag` (see also #223). I'm also wondering now whether there is any reason to even keep `vA`, or whether using the linear transpose should _always_ be used. Do you have intuition for this? (How do you feel about opening a separate issue for this, so we can discuss both features separately?)

As for the second one, I understand that for CG this is super convenient. For the SVD I am a bit less sure actually, because U and V would then also become pytrees and working with the SVD might become relatively complicated. Do you have a concrete use-case where pytree-valued SVDs would be helpful?  

>      imshow(comps[0][i])

Just to clarify: Are you sure this line shows the $i$th singular vector instead of showing the $i$th component  of all singular vectors?

Looking forward to hearing your thoughts!

**Comment by tjltjl** [comment]
I split 1. into #228 .

Re your question: the shape of the comps and the resulting image do make it clear that it is the $i$th singular vector.

The choice here is basically about which end of the tensor axes to place the new axis in, beginning (as in here) or the end. In the example code, I chose to place the singular vector index as the 0th axis (by using jax.vmap(u_unravel)(u.T)). I think it would make sense to make the same choice for both u and v, but that could be customized by an "out_axis" argument with default (-1, 0); setting it to 0 would get the behaviour of my example function above.

For this particular example, the pytreeness does not make things too difficult since the tree is just a 2D tensor instead of a 1D tensor.

The simplest concrete example is basically the above case: we have some structure in the basis. Could be a 2D discretized array, or could be a multiphysics system with two separate FEM bases. Not having to transform to a structureless vector makes things that much simpler both outside the call and inside the Av op.

(doing hermitian eigenproblem and selecting the smallest/largest/value near given value instead of svd would actually be even better for some cases where A^T A v would be easier to evaluate if Av is a longer vector)

**Comment by pnkraemer** [comment]
How about something like the below?

The only difference to your version is that I replaced the nonlocal variable with liberal usage of `jax.eval_shape` (note that this requires wrapping the output of jax.flatten_util.ravel_pytree through a jax.tree_util.Partial). 

```python

# https://github.com/pnkraemer/matfree/issues/226
def svd_partial_simple(v0, depth, Av):
    # Evaluate the output shape of Av and flatten
    u0 = jax.eval_shape(Av, v0)

    # Flatten in- and outputs
    _, u_unravel = jax.eval_shape(_ravel_pytree, u0)
    v0_flat, v_unravel = _ravel_pytree(v0)

    def Av_flat(v_flat):
        """Evaluate a flattened matvec."""
        result = Av(v_unravel(v_flat))
        result_flat, _ = _ravel_pytree(result)
        return result_flat

    def vA_flat(u_flat):
        """Evaluate a flattened vecmat."""
        (v_flat,) = jax.linear_transpose(Av_flat, v0_flat)(u_flat)
        return v_flat

    # Call the flattened SVD
    u, s, v = eig.svd_partial(v0_flat, depth, Av_flat, vA_flat)

    # Select out_axes to ensure consistency with U @ diag(s) @ Vh
    u_tree = jax.vmap(u_unravel, out_axes=-1)(u.T)
    v_tree = jax.vmap(v_unravel, out_axes=0)(v)
    return u_tree, s, v_tree


def _ravel_pytree(tree):
    # Wrap the `unravel` function through a Partial
    # to enable jax.eval_shape
    flat, unravel = jax.flatten_util.ravel_pytree(tree)
    return flat, jax.tree_util.Partial(unravel)
```

Thoughts?

**Comment by tjltjl** [comment]
Nice! I didn't realize you could do that with jax.eval_shape + ravel.

We'll have to agree to disagree about the output axis - to me, once you upgrade from matrices to tensors, the linear ordering loses its meaning and it's better to be consistent.

But I do understand that this is the convention.

Re 2D: that was the simplest form of the example I could make; in reality, the 2D array is inside an equinox module that contains other parameters about how exactly to interpret that array into a function.

Also, I imagine that for some problems it would be cool to use this on a jacobian of a neural network; at that point, the pytree shape gets quite involved :)

**Comment by pnkraemer** [comment]
#236 implements the partial SVD as sketched above. Notably, I decided to make the "singular-value index" the leading axis of all decompositions -- you convinced me. Beyond your reasons, I realised it's also easier to document, and easier to vmap over the outputs of the SVD if all arrays have the same leading dimension.


I'll follow up with similar changes for eig_partial and eigh_partial and draft a new release (v0.2.1) afterwards. Thanks for the discussion, I believe this enhancement improves Matfree's API! Let me know if you have more suggestions :)

**Comment by tjltjl** [comment]
This is awesome, I'll let you know once I have a moment to test this. This is the go-to linalg tool I want to have :)
